### DISCUSSION TITLE
[Idea] Forbidden php access on modules directory

### DISCUSSION THREAD

**Discussion by clotairer** [PrestaShop/PrestaShop#32566] [Ideas]
### The problem

Since PrestaShop 1.5, ModuleFrontController and ModuleAdminController exists and let developers to manage Front and Admin route to manager custom pages.

Too often we found on modules code like that :

```php
# cron.php or ajax.php or api.php
<?php
// modules/mymodule/ajax.php
include(dirname(__FILE__) . '/../../config/config.inc.php');
include(dirname(__FILE__) . '/../../init.php');
$mymodule = new mymodule();

echo $mymodule->displayAjax();
```

There are too many weakness associated to this:
- This kind of code often use a poor protection like `Tools::hash(false)` that is a predictable hash.
- Mosts modules forget to add a condition that check if the module is enabled.
- I also see in a partner module an api endpoint that enable the module `$mymodule->enable()` with a simple  HTTP call ! This kind of code don't load the core dispatcher and all core protections like authentication.

This is already register as a bad practice in the [PrestaShop Dev Docs](https://devdocs.prestashop-project.org/1.7/modules/creation/good-practices/) :
"Don’t go through directories using code and variables like: dirname(__FILE__).'/../../config/config.inc.php'"

### The solution

We should consider for next major PrestaShop release (9) that work around the core `Dispatcher` is a vulnerability.

That's why, I propose to add in the modules/.htaccess file a **deny** directive for all *.php files extensions. This is a simple way to encourage developers to use ModuleController instead.

More over, I add an important security improvement. To be catched easily by WAF:
- `fc`, `module` and `controller` parameters should be send only with a GET method (not in POST).
- If Rewrite Rules is enabled, `fc`, `module` and `controller` should be ignored. The rewrite route should only be used.

### Additional context

More than one out of two critical [vulnerabilities disclosed by Friends Of Presta](https://friends-of-presta.github.io/security-advisories/), involve this kind of legacy pseudo controller. In most cases if the module is not installed the vulnerability remains exploitable. That’s not fair a disabled module can have security impact on a shop !

### Do you plan to work on this subject?

- [X] I'm willing to contribute a [formal specification](https://github.com/PrestaShop/prestashop-specs).
- [X] I'm willing to provide any wireframes or design assets required for this feature.
- [X] I'm willing to submit a Pull Request that implements this feature.
- [X] I'm willing to help verify that the implemented feature works as intended and produces no unintended side effects.

**Comment by matks** [comment]
> That's why, I propose to add in the modules/.htaccess file a deny directive for all *.php files extensions. This is a simple way to encourage developers to use ModuleController instead.

Hohoho 😄 that is the kind of move where we have no idea if we're going to get a tsunami of praises or a tsunami of hate ...

I like the idea, but I fear the reaction of the community. I have no idea if we can do this, but if we do, we have to communicate early and multiple times on this change 😝

**Reply by ghost** [reply]
As for the php version drop, you can make a blog post

**Comment by clotairer** [comment]
@matks, you are right by sharing us your feeling but facts are here. 
More than one out of two critical [vulnerabilities disclosed by Friends Of Presta](https://friends-of-presta.github.io/security-advisories/), involve this kind of legacy pseudo controller. In most cases if the module is not installed the vulnerability remains exploitable. That’s not fair that a disabled module can have security impact on a shop !
(I added this crucial argument in the main discussion)

**Comment by touchweb-vincent** [comment]
> More over, I add an important security improvement. To be catched easily by WAF:
> 
> * `fc`, `module` and `controller` parameters should be send only with a GET method (not in POST).
> * If Rewrite Rules is enabled, `fc`, `module` and `controller` should be ignored. The rewrite route should only be used.


:) +1

**Comment by ghost** [comment]
I totally agree and to join [this discussion](https://github.com/PrestaShop/PrestaShop/discussions/32358) since you are sure that module editors will make the move to drop php < 8.1

![Screenshot (9)](https://github.com/PrestaShop/PrestaShop/assets/5064590/e59243ce-7e17-4ab8-9006-dc410d5f7d28)

**Comment by boherm** [comment]
I like this idea but we need to be carefull. 
Adding a rule on `.htaccess` can improve security but It can be rude for module editors.
However, we are performing a big step to drop old php versions on PrestaShop 9, so ... It can be the good moment to add this thing.

**Comment by matks** [comment]
> However, we are performing a big step to drop old php versions on PrestaShop 9, so ... It can be the good moment to add this thing.

Warning 😛 : I see a lot of comments "you decided to drop old PHP versions which is going to anger people, so let's continue pushing unpopular but useful decisions, now that we started"

Actually this is a bad idea 😄 : if we put TOO MUCH unpopular decisions in one version we have a high risk that this version is not welcomed 😉 

So it's a matter of balance. IIn my opinion in PrestaShop 9 we have to carefully put _a fair number of unpopular yet important changes_ to allow the software to move forward, but _not too much_.

If we make PrestaShop 9 too unpopular, we'll have the reverse effect. By strengthening PrestaShop 9 we hope that _people that will migrate to PS9 will be more secure_. But if people do NOT migrate to PS9 because we've put too much constraints in it, and stay on older versions, we achieve the opposite. If nobody uses PS9 because it introduces too many big changes, we failed. We win the game if we successfully convince people to upgrade to PS9 and benefit from the higher security, although upgrading to PS9 will have a price.

So ... it's about the right number of changes. Not zero, not a thousand. The right quantity 😉 .

**Comment by ghost** [comment]
@matks 

_We win the game if we can convince people to upgrade to PS9 and get the added security, even if upgrading to PS9 will come at a price._

If I follow your reasoning, it's better to add .htaccess than to abandon older versions of php for the PrestaShop ecosystem.
Because hacks come from these pseudo controllers but not from the version of php that will help more the devs
We can remove 80% of hacks with a small file.

**Comment by jf-viguier** [comment]
@matks using a cron/ajax without a controller has been discouraged since prestashop 1.7.6 I guess, so it's not a big move IMO for module developers.
For custom development that can't be changed, it's also easy to remove this line in the htaccess.

**Comment by PrestaEdit** [comment]
About upgrade and dropping PHP support or not, you need to know that even PrestaShop 8 introduced a large needs to handle an update on a large set of modules.

Even if the published articles before it say that : "It's not a big deal, its like a PrestaShop 1.7 without Addons related [and you will be able to upgrade without worries]", it was not the case.
A lot of modules was using the `Tools::jsonEncode` method and it was dropped. Okay, this method was deprecated since 1.7 version but whoes know ?

Now, it's another game.
There is a usefull documentation on notables changes and it's more easier to understand the changes.

A lot of modules doesn't includes theses scripts inside them. Because : 
- they know that it's not secure ;
- this is not more readable/easier to use ;
- it's a "Don't do" part of the technical documentation since years ;
- it's a reason to be not published on the Addons Marketplace

Even on Back Office side, you can use your own controller if it's already existing and/or rely on the AdminModules one that can catch them for your module and dispatch them to it.

If you think that this kind of changes will be the main reason to not upgrade an existing store, I think you don't understand how merchants and developpers works with PrestaShop.
They will make the move especially because there is a real interest behind it, not because it's change the version or something as it, for sure.

**Comment by matks** [comment]
Ah 😄 finally looks like we're able to have detailed and reasoned discussion ! I'm happy to see this 😊

**Comment by kpodemski** [comment]
I like the idea. Let's make PrestaShop more secure. We'd communicate this as soon as possible.

**Comment by kpodemski** [comment]
I can think of one downside. Sometimes there are modules adding Orders, Customers, etc. mostly, integrators with some marketplaces, or import modules, where if you use `ModuleFrontController`, you may have many issues with `Context`.

I think I'll reach out to some module developers to ask if the migration to `ModuleFrontController` is possible. (one option could be to add some allow-list 🤔 )

**Reply by clotairer** [reply]
As developer of Shoppingfeed marketplace module or payment modules, we found solutions to manage front controller in a « all shops context » at all times.
Perhaps you should verify on ps_facetedsearch why you have ps_facetedsearch-attribute-indexer.php (and others) and not a front controller instead ? You probably found a response by requesting a developper of this module.

**Reply by eternoendless** [reply]
I think it's just old code that was never updated

**Comment by eternoendless** [comment]
> I propose to add in the modules/.htaccess file a deny directive for all *.php files extensions. This is a simple way to encourage developers to use ModuleController instead.

Perhaps it could be made as an option that people can turn off for compatibility.

**Reply by clotairer** [reply]
I don't think it's a good idea. If you do that you need to manage exceptions module by module.
I think we need to assume this breaking changes.

We developped for our merchand a module to manage allowed module with PHP scripts, and allowed extensions. Because, there is a risk of data leak too with .csv .log .txt ... But it's awefull to manage these necessary kind of rules. Modules developpers need to manage properly rights to read files and use front controllers without by passing PrestaShop Dispatcher that warranty to exec Hooks and all PrestaShop core security.

If you have a security issue in a modules/mymodule/ajax.php script, you can't do nothing to stop an attack. If you have a call like /module/mymodule/ajax with the same vulnerability, an other security module (a WAF module) can stop an attack with a hook on actionFrontControllerInitBefore for instance.

**Reply by jf-viguier** [reply]
I totally agree with you @clotairer but keep in mind that modules can save images or css in their folders, so maybe the htaccess rule should allow "safe" extensions.

**Reply by clotairer** [reply]
static files should be allowed (same htacces than themes/.htaccess)

**Comment by nicosomb** [comment]
@PrestaShop/tech-council Could you please give your opinion about this discussion please?

**Reply by eternoendless** [reply]
I like the idea but I'm just worried that modules might stop working because of the hard block. That's why I suggested making it "disableable", ideally per module, with a warning.

**Reply by jf-viguier** [reply]
If we do that, rotten modules will put in their docs to add their module in the exclusions and there will still be php files without controllers in 10 years, with security flaws.
I'd suggest to add a full blocking as an experimental feature and to warn that the next major will disallow php files without controller.
 php files without controller is not allowed in the doc since 1.7 I guess.
 
 What legit use case do you see where a module could launch a php file without controller ?

**Reply by eternoendless** [reply]
From an engineering perspective I agree with you @jf-viguier. But from a user perspective I wouldn't want to block a user from using their "insecure" module, as long as they make an informed choice.

**Reply by clotairer** [reply]
He can modify the .htacces by himself to remove le deny policy

**Reply by jolelievre** [reply]
After reading this all I also agree it would be a good idea, at least the htaccess part The part about parameters in GET/POST doesn't seem related and I don't understand what it's a problem but maybe it should be dealt with in another discussion

And regarding the possibility of disabling this I agree that if it's possible people will still use the opportunity not to do things correctly, if we want to steadily improve the quality of the core that's some of the decisions we should take upon ourselves even if it can be unpopular for SOME modules' developers

And I insist on the SOME, because I actually ould be curious to know how many modules still use this kind of bad practice, I don't know if we can have some data about this from Addons? Because maybe there are not that many and we're actually complexifying this decision that would actually have an impact on a limited number of modules

Finally, I just discovered this PR https://github.com/PrestaShop/PrestaShop/pull/33700 and I think it's kind of linked to this topic, what do you guys think? Just like for this discussion I kind of lack vision about how this could impact the ecosystem

**Reply by clotairer** [reply]
#33700 is the same EPIC but not the same issue

About POSTed fc module and controller parameters, I can open an independant discussion to explain the SEO (several routes to the same page) impact on the security (especially for Forensic attack analysis).

**Reply by MatShir** [reply]
For Addons POV, it is forbidden since this year.
- https://docs.cloud.prestashop.com/10-validation-checklist/#an-htaccess-file-exists-in-the-root-folder-of-the-module
- https://docs.cloud.prestashop.com/10-validation-checklist/#php-files-are-executed-in-prestashop-context

**Reply by Thymotep** [reply]
Hello everyone,

As of this year, we have [introduced a rule](https://docs.cloud.prestashop.com/10-validation-checklist/#an-htaccess-file-exists-in-the-root-folder-of-the-module) on this subject for modules submitted to the PrestaShop marketplace.

In the coming weeks, the presence of the file will be checked automatically when validating a module (new or update) and the module will be refused if the file is not present.

So I have no objection to it being integrated into the core, as long as it only concerns .php files (image assets, js etc can be accessed from the front-office).

**Comment by matks** [comment]
The PR https://github.com/PrestaShop/PrestaShop/pull/34184 was merged consequently I close this discussion

**Comment by matks** [comment]
I have created a ticket to follow the next steps needed https://github.com/PrestaShop/PrestaShop/issues/34578 for 9.0.0 release

I hope the people who participated in this discussion to advocate for this change will be helping this task to be complete 😝 . I plan to submit a few PRs to fix some of the modules listed but I will not do all of them.
